sdks/python/apache_beam/io/filesystemio.py

```diff
@@ -0,0 +1,492 @@
+    """
+    return self._read_inner(size=size, readline=True)
+
+  def _read_inner(self, size=-1, readline=False):
+    """Shared implementation of read() and readline()."""
+    self._check_open()
+    if not self._remaining():
+      return ''
+
+    # Prepare to read.
+    data_list = []
+    if size is None:
+      size = -1
+    to_read = min(size, self._remaining())
+    if to_read < 0:
+      to_read = self._remaining()
+    break_after = False
+
+    while to_read > 0:
+      # If we have exhausted the buffer, get the next segment.
+      # TODO(ccy): We should consider prefetching the next block in another
+      # thread.
+      self._fetch_next_if_buffer_exhausted()
+
+      # Determine number of bytes to read from buffer.
+      buffer_bytes_read = self.position - self.buffer_start_position
+      bytes_to_read_from_buffer = min(
+          len(self.buffer) - buffer_bytes_read, to_read)
+
+      # If readline is set, we only want to read up to and including the next
+      # newline character.
+      if readline:
+        next_newline_position = self.buffer.find('\n', buffer_bytes_read,
+                                                 len(self.buffer))
+        if next_newline_position != -1:
+          bytes_to_read_from_buffer = (
+              1 + next_newline_position - buffer_bytes_read)
+          break_after = True
+
+      # Read bytes.
+      data_list.append(self.buffer[buffer_bytes_read:buffer_bytes_read +
+                                   bytes_to_read_from_buffer])
+      self.position += bytes_to_read_from_buffer
+      to_read -= bytes_to_read_from_buffer
+
+      if break_after:
+        break
+
+    return ''.join(data_list)
+
+  def _fetch_next_if_buffer_exhausted(self):
+    if not self.buffer or (
+        self.buffer_start_position + len(self.buffer) <= self.position):
+      bytes_to_request = min(self._remaining(), self.buffer_size)
+      self.buffer_start_position = self.position
+      try:
+        result = self._get_segment(self.position, bytes_to_request)
+      except Exception as e:  # pylint: disable=broad-except
+        tb = traceback.format_exc()
+        logging.error(
+            ('Exception while fetching %d bytes from position %d of %s: '
+             '%s\n%s'),
+            bytes_to_request, self.position, self.path, e, tb)
+        raise
+
+      self.buffer = result
+      return
+
+  def _remaining(self):
+    return self.downloader.size - self.position
+
+  def close(self):
+    """Close the current file."""
+    self.closed = True
+    self.download_stream = None
+    self.downloader = None
+    self.buffer = None
+
+  def _get_segment(self, start, size):
+    """Get the given segment of the current file."""
+    if size == 0:
+      return ''
+    end = start + size - 1
+    self.downloader.get_range(start, end)
+    value = self.download_stream.getvalue()
+    # Clear the cStringIO object after we've read its contents.
+    self.download_stream.truncate(0)
```

Comment:
<!--new_thread; commit:970afe0523d6d5673489a70e801be6553e0ffc33; resolved:0-->
It looks like the reason we have a stream at all is to satisfy the particular API of the apitools client we use for GCS.  Would a cleaner API be to have the `downloader` directly return the bytes?  (so that the `download_stream` would be managed by the downloader?)

See above for how we could get rid of `Downloader.start(download_stream, ...)` with such an approach.

Reply:
get_range now returns a string.